sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SasTokenProviderWithSharedAccessKey.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Azure.Iot.Hub.Service.Authentication
+{
+    public class SasTokenProviderWithSharedAccessKey : ISasTokenProvider
+    {
+        private static readonly TimeSpan s_defaultTimeToLive = TimeSpan.FromMinutes(30);
+        private readonly object _lock = new object();
+
+        private readonly string _hostname;
+        private readonly string _sharedAccessPolicy;
+        private readonly string _sharedAccessKey;
+        private readonly TimeSpan _timeToLive;
+
+        private string _cachedSasToken;
+        private DateTimeOffset _tokenExpiryTime;
+
+        // Private constructor to prevent accidental initialzation.
+        private SasTokenProviderWithSharedAccessKey()
+        {
+        }
+
+        public SasTokenProviderWithSharedAccessKey(string hostname, string sharedAccessPolicy, string sharedAccessKey, TimeSpan? timeToLive = null)
+        {
+            _hostname = hostname;
+            _sharedAccessPolicy = sharedAccessPolicy;
+            _sharedAccessKey = sharedAccessKey;
+            _timeToLive = (TimeSpan)(timeToLive == null ? s_defaultTimeToLive : timeToLive);
+
+            _cachedSasToken = null;
+        }
+
+        public string GetSasToken()
+        {
+            lock (_lock)
+            {
+                if (IsTokenExpired())
```

Comment:
I'm a bit worried about clock skew causing the SDK to send what it thinks is still a valid token only for the service to reject it. What do you think about a slightly more proactive renewal? Something like at 85% of the sas token's lifespan instead of 100%?

Reply:
Sure, that's doable.

Reply:
Did we conclude to do this and is there a task to track?

Reply:
It is already implemented.